### ISSUE TITLE
multiple definition of 'setUp' and 'tearDown'

### ISSUE THREAD

**Issue by ecorini** [ThrowTheSwitch/Ceedling#651]
Hi, 

Sometimes in test modules i have errors like this:
```
test_MultipleDefinitionError.c:7: multiple definition of `setUp'; build/test/out/c/test_MultipleDefinitionError_runner.o:C:\Progetti\Temp\MultipleDefinitionError/build/test/runners/test_MultipleDefinitionError_runner.c:30: first defined here
```
The generated runner file contains `setUp` and `tearDown` methods also if they are present in the test module.

I tried to reproduce it with a minimal example

[MultipleDefinitionError.zip](https://github.com/ThrowTheSwitch/Ceedling/files/7805946/MultipleDefinitionError.zip)

The problem seems related on #ifdef statements, and disappear if i remove   `:use_preprocessor_directives: TRUE` in project.yml.
The problem also disappear if i remove empty line 23 from file `test_MultipleDefinitionError.c`.

I'm on a windows platform with GCC on MSYS2:

```
gcc (Rev1, Built by MSYS2 project) 11.2.0

Ceedling:: 0.31.1
CMock:: 2.5.4
Unity:: 2.5.4
CException:: 1.3.3
```

**Comment by mundodisco8** [comment]
I'm having the same issue. I need the preprocessor for my tests, but with it enabled, as soon as I have a `#define SOMETHING` before the `setUp()` function, the linker complains of it being defined multiple times... The issue disappear 

I'm running v1.0

```
Welcome to Ceedling!

  Ceedling => 1.0.0-f08576b
  ----------------------
  C:/Ruby33-x64/lib/ruby/gems/3.3.0/gems/ceedling-1.0.0/

  Build Frameworks
  ----------------------
       CMock => 2.6.0
       Unity => 2.6.0
  CException => 1.3.4
```

My workaround, for now, has been to change `:use_test_preprocessor:` from `:all` to `:`mocks`

**Comment by ForestRupicolous** [comment]
I have the same issue trying the example from [ceedling](https://www.throwtheswitch.org/ceedling) 
`Welcome to Ceedling!

  Ceedling => 1.0.0
  ----------------------
  C:/Tools/Ruby30-x64/lib/ruby/gems/3.0.0/gems/ceedling-1.0.0/

  Build Frameworks
  ----------------------
       CMock => 2.6.0
       Unity => 2.6.1
  CException => 1.3.4`
and the workaround is not helping.

**Comment by mvandervoord** [comment]
Hi @ForestRupicolous  and @mundodisco8.

If you are able, can you send us more information? It'd be nice to have an example of a file with this problem. I'm not currently able to reproduce it using the website's sample project, nor by just adding a `#define` into an existing file. :/

**Comment by ForestRupicolous** [comment]
Hi @mvandervoord ,
thank you for your fast reply. 
I'm working under Win10.
I freshly updated the ceedling gem via `gem install ceedling`
and the just followed the steps on the webpage:
Create new project, change into folder, create module, copy code into Demo.c, run all tests.
I just start with using ceedling but it looks like the linke is unhappy that test_Demos.c and Demos.c both contain setUp and tearDown functions.
Additionally I updated my Ruby installation as I expected an issue with the contained gcc-toolchain.
The files:
``` c
// Demo.c
#include "Demo.h"

int run_demo(int a, int b)
{
  return 2 * a + b;
}

// Demo.h
int run_demo(int a, int b);

// test_Demo.c
#include "unity.h"
#include "Demo.h"

void setUp(void) {}
void tearDown(void) {}

void test_RunningDemo_should_DoNormalOperations(void)
{
  TEST_ASSERT_EQUAL_INT(3, run_demo(1,1));
  TEST_ASSERT_EQUAL_INT(6, run_demo(2,2));
  TEST_ASSERT_EQUAL_INT(13, run_demo(4,5));
}
```
test_Demos.c (as generated)
```
#ifdef TEST

#include "unity.h"

#include "Demo.h"

void setUp(void)
{
}

void tearDown(void)
{
}

void test_Demo_NeedToImplement(void)
{
    TEST_IGNORE_MESSAGE("Need to Implement Demo");
}

#endif // TEST
```

**Comment by Letme** [comment]
Who generated test_demo.c? Why does demo.c contain setUp/tearDown?

**Comment by mundodisco8** [comment]
@Letme I'm not the original author, but I think he just placed all the code in the same code block, notice the comments with `// demo.c`, `// demo.h` and `// test_demo.c`

**Comment by Letme** [comment]
Yes, but the generated test_demo.c is in separate code block, and has a new test (so maybe test_demo2.c? But anyway it writes generated, so I am wondering generated by what. And he wrote that linker is unhappy because test_demo.c and demo.c have setUp and tearDown - which means there is a setUp in demo.c (which is not in code block example). I am just trying to get some clarity

**Comment by mundodisco8** [comment]
You are right. I don't want to speak for @ForestRupicolous , so I'll shut up 😄 

Regarding my report, I updated to v1.0 proper (the current release) and I can't replicate either, so I would say my issue was fixed somewhere between f08576b and the release.

**Comment by ForestRupicolous** [comment]
Hi,
have you tried following the instructions on the linked [website](https://www.throwtheswitch.org/ceedling)?
When a new module is created ('ceedling module:create[Demo]) the file test_Demo.c is automatically generated (additionally to an empty Demo.c and Demo.h). When I then copy the Demo.c code from the website I get the linker errors. 
I"m just starting with ceedling and from your messages it looks like there is a misunderstanding of how the files should be structured and the Demo.c code provided on the website.

**Comment by ForestRupicolous** [comment]
Yes, this embarrassing oversight created my issue. It works now and make more sense.  
Thank you all again for the fast response and explanation.

**Comment by Letme** [comment]
Ok, so core issue is that those files on website need to go in separate code blocks as they cause confusion when they are only separated by comments.

**Comment by mvandervoord** [comment]
@Letme -- agreed! I'll fix that. Sorry about the confusion, everyone!

**Comment by mvandervoord** [comment]
https://throw-the-switch.squarespace.com/config/pages

I believe this is more clear now? Thanks for the help everyone!

**Comment by Letme** [comment]
https://throw-the-switch.squarespace.com/ceedling (your other link points to admin pages).

Looks as expected. Thanks for the clarification ;)
